doc/SAI-Proposal-Debug-Counters.md

```diff
@@ -0,0 +1,239 @@
+Created     | 07/17/2019
+SAI-Version | 1.5
+
+## Overview
+ASICs can provide a set of debug counters for certain object types.
+Debug counters belong to certain families, each family is for cetain object type.
+The content of a specific debug counter instance can be defined by the application.
+Counting every statistics of every family might be too resource intensive, therefore the debug counters provide an efficient way to count and aggregate only the needed information.
+
+## Debug counter object type
+For the counter object, a sai_counter_stat_t enum is defined:
+```
+typedef enum _sai_debug_counter_type_t
+{
+    /** Port in drop reasons. Base object : SAI_OBJECT_TYPE_PORT */
+    SAI_DEBUG_COUNTER_TYPE_PORT_IN_DROP_REASONS,
+
+    /** Port out drop reasons. Base object : SAI_OBJECT_TYPE_PORT */
+    SAI_DEBUG_COUNTER_TYPE_PORT_OUT_DROP_REASONS,
+
+} sai_debug_counter_type_t;
+```
+
+## Debug counter port in drop reason family
+```
+typedef enum _sai_port_in_drop_reason_t
+{
+    /* L2 reasons */
+	
+    /** Source MAC is multicast */
+    SAI_PORT_IN_DROP_REASON_SMAC_MULTICAST,
+
+    /** Source MAC equals Destination MAC */
+    SAI_PORT_IN_DROP_REASON_SMAC_EQUALS_DMAC,
+
+    /** Destination MAC is Reserved (DMAC=01-80-C2-00-00-0x) */
+    SAI_PORT_IN_DROP_REASON_DMAC_RESERVED,
+
+    /** 
+     * @brief VLAN tag not allowed
+     *
+     * Frame tagged when port is dropping tagged, 
+     * or untagged when dropping untagged
+     */
+    SAI_PORT_IN_DROP_REASON_VLAN_TAG_NOT_ALLOWED,
+
+    /** Ingress VLAN filter */
+    SAI_PORT_IN_DROP_REASON_INGRESS_VLAN_FILTER,
+
+    /** Ingress STP filter */
+    SAI_PORT_IN_DROP_REASON_INGRESS_STP_FILTER,
+	
+    /** Unicast FDB table action discard */
+    SAI_PORT_IN_DROP_REASON_FDB_UC_DISCARD,
+
+    /** Multicast FDB table empty tx list */
+    SAI_PORT_IN_DROP_REASON_FDB_MC_DISCARD,
+
+    /** Port loopback filter */
+    SAI_PORT_IN_DROP_REASON_LOOPBACK_FILTER,
+
+    /* L3 reasons */	
+
+} sai_port_in_drop_reason_t;
+```
+
+## Debug counter port out drop reason family
+```
+typedef enum _sai_port_out_drop_reason_t
+{
+    /** Egress VLAN filter */
+    SAI_PORT_OUT_DROP_REASON_EGRESS_VLAN_FILTER,
+
+} sai_port_out_drop_reason_t;
+```
+
+### Configuring a debug counter
+```
+/**
+ * @brief Attribute Id in sai_set_counter_attribute() and
+ * sai_get_counter_attribute() calls
+ */
+typedef enum _sai_debug_counter_attr_t
+{
+    /**
+     * @brief Start of attributes
+     */
+    SAI_DEBUG_COUNTER_ATTR_START,
+
+    /* READ-WRITE */
+
+    /**
+     * @brief Debug counter type
+     *
+     * @type sai_debug_counter_type_t
+     * @flags MANDATORY_ON_CREATE | CREATE_ONLY
+     * @isresourcetype true
+     */
+    SAI_DEBUG_COUNTER_ATTR_TYPE = SAI_DEBUG_COUNTER_ATTR_START,
+
+    /**
+     * @brief Object stat index
+     * Index is added to base start
+     *
+     * @type sai_uint32_t
+     * @flags MANDATORY_ON_CREATE | CREATE_ONLY
+     */
+    SAI_DEBUG_COUNTER_ATTR_INDEX,
```

Comment:
As discussed in the community meeting, can you please consider making this a READ_ONLY and let SAI manage the indexing ?

Reply:
+1, I think letting the SAI manage the indexing would be less prone to user error

Reply:
Changed to RO as requested